sys/kern/rman.c

```diff
@@ -166,6 +182,9 @@ void rman_release_resource(resource_t *r) {
   assert(!r_active(r));
   assert(r_reserved(r));
 
+  if (r_shareable(r) && --r->r_refcnt)
+    return;
```

Comment:
I think you might be abusing `rm_lock` to protect the reference counter. If it was only the counter that was shared between resource sharers, each sharer could hold its own `resource_t`, fields such as `rid` or `handler` wouldn't have to be pushed out of `resource_t` to `resource_list_entry_t` and decrementing `r_refcnt` wouldn't block the entire rman.

To be clear - I'm not 100% sure that this can be easily solved by adding a separate lock on `r_refcnt`, but it it can be solved that way, I'd be much more in favor of such solution.

Reply:
Ok, I just realized that in the previous solution share lists were present for this very reason to allow devices to hold their own `resource_t`s. They ensured that after releasing a resource rman would still hold some `resource_t` to mark the physical resource as claimed. Going back to having multiple `resource_t`s for one shared resource would be in fact going back to the previous solution.

After thinking about it for a while, I decided that this new approach is indeed much better after all! I just reimagined `resource_list_entry_t` to be named as something like `resource_handle_t` and it all made sense! There's a clear dichotomy ie. `resource_t` holds the data that defines the resource along with user count, and `resource_list_entry_t` hold any auxiliary data that a driver ties to the resource. I wonder if we could use `resource_list_entry_t`s as handles to resources iinstead of `resource_t *`s in drivers and just get rid of `resource_rle`. This feels like a natural step after moving all that auxiliary data to `resource_list_entry_t`.

However, all this weghing doesn't really affect my point about abusing `rm_lock`. Correct me, if I'm wrong, but I'm pretty sure that it would be a good idead to introduce a per-resource lock that would protect the reference counter. without locking the entire rman if not necessary.